storage/range.go

```diff
@@ -782,52 +782,51 @@ func (r *Range) applyRaftCommand(index uint64, originNodeID multiraft.NodeID, ar
+		log.Fatalf("applied index moved backwards: %d >= %d", oldIndex, index)
+	}
+
+	// Advance the applied index atomically within the batch.
+	if e := setAppliedIndex(batch, r.Desc().RaftID, index); e != nil {
+		if reply.Header().GoError() == nil {
+			reply.Header().SetGoError(e)
 		}
+		err = e
+	}
 
-		if proto.IsWrite(args) {
-			// On success, flush the MVCC stats to the batch and commit.
-			r.stats.MergeMVCCStats(batch, &ms, header.Timestamp.WallTime)
-			if err := batch.Commit(); err != nil {
-				log.Fatalf("failed to commit batch from Raft command execution: %s", err)
-			} else {
-				// After successful commit, update cached stats values.
-				r.stats.Update(ms)
-				// Publish update to event feed.
-				r.rm.EventFeed().updateRange(r, args.Method(), &ms)
-				// If the commit succeeded, potentially add range to split queue.
-				r.maybeSplit()
-				// Maybe update gossip configs on a put.
-				switch args.(type) {
-				case *proto.PutRequest, *proto.DeleteRequest, *proto.DeleteRangeRequest:
-					if header.Key.Less(engine.KeySystemMax) {
-						r.maybeGossipConfigs(func(configPrefix proto.Key) bool {
-							return bytes.HasPrefix(header.Key, configPrefix)
-						})
-					}
-				}
+	if err == nil && proto.IsWrite(args) {
+		// On success, flush the MVCC stats to the batch and commit.
+		r.stats.MergeMVCCStats(batch, &ms, header.Timestamp.WallTime)
+		if err := batch.Commit(); err != nil {
+			log.Fatalf("failed to commit batch from Raft command execution: %s", err)
+		}
+		// After successful commit, update cached stats and appliedIndex value.
+		atomic.StoreUint64(&r.appliedIndex, index)
+		r.stats.Update(ms)
+		// Publish update to event feed.
+		r.rm.EventFeed().updateRange(r, args.Method(), &ms)
+		// If the commit succeeded, potentially add range to split queue.
+		r.maybeSplit()
+		// Maybe update gossip configs on a put.
+		switch args.(type) {
+		case *proto.PutRequest, *proto.DeleteRequest, *proto.DeleteRangeRequest:
+			if header.Key.Less(engine.KeySystemMax) {
+				r.maybeGossipConfigs(func(configPrefix proto.Key) bool {
+					return bytes.HasPrefix(header.Key, configPrefix)
+				})
 			}
 		}
 	} else {
-		atomic.StoreUint64(&r.appliedIndex, index)
-		// On failure, abandon the batch we've built up, but still update the
-		// applied index so we won't retry this command on restart.
-		if err := engine.MVCCPut(r.rm.Engine(), nil, engine.RaftAppliedIndexKey(r.Desc().RaftID),
-			proto.ZeroTimestamp, proto.Value{Bytes: encoding.EncodeUint64(nil, index)}, nil); err != nil {
-			// The reply header already contains an error which is going to be more useful
-			// to the caller than this one, so just log it.
-			log.Errorf("failed to advance applied index: %s", err)
+		// Advance the last applied index nonetheless, without the batch.
+		if err = setAppliedIndex(r.rm.Engine(), r.Desc().RaftID, index); err != nil {
+			log.Errorf("could not advance applied index: %s", err)
```

Comment:
This could cause replicas to diverge; it should probably be a Fatalf just like the one on line 804.